KIPs/kip-146.md

```diff
@@ -0,0 +1,96 @@
+---
+kip: 146
+title: Random Proposer Selection
+author: Ian (@ian0371), Ollie (@blukat29), Joseph (@jiseongnoh), and Aidan (@aidan-kwon)
+discussions-to: https://github.com/klaytn/kips/issues/146
+status: Draft
+type: Standards Track
+category: Core
+created: 2023-06-15
+---
+
+## Simple Summary
+
+An unpredictable and verifiable proposer selection policy
+
+## Abstract
+
+This standard outlines an approach for a random proposer selection policy. A proposer is determined every block in a random manner. The randomness relies on [KIP-114](https://github.com/klaytn/kips/blob/main/KIPs/kip-114.md), thereby rendering the new policy both unpredictable and verifiable.
+
+## Motivation
+
+Klaytn has adopted the Byzantine Fault Tolerance (BFT) consensus mechanism, where a block proposer selected by a deterministic algorithm proposes the next block. A vulnerable point of BFT is that an adversary can effectively halt the entire network by simply targeting the proposer, a.k.a. targeted DoS. What makes Klaytn even more vulnerable to a targeted DoS attack is that the current proposer selection policy enables the prediction of up to `proposerRefreshInterval` (3600 in case of Cypress) number of proposers in advance.
```

Comment:
> the current proposer selection policy enables the prediction of up to `proposerRefreshInterval` (3600 in case of Cypress) number of proposers in advance.

To my knwoledge, proposer list would not be updated until 1 day pass over in Cypress network. (https://github.com/klaytn/klaytn/blob/dev/consensus/istanbul/validator/weighted.go#L643)  Thus, it wouldn't trigger the shuffling process against the current proposers before 1 day.


A terminology `proposerUpdateInterval` seems better than `proposerRefreshInterval` as the former is used in `genesis.json`, but the latter appears only in the source code.

Reply:
I think "no staking info" is a rare error case.

Usually, `blockNum` is not the latest block, it is the block number of the output `CalcProposerBlockNumber`. The argument passing occurs here: 
https://github.com/klaytn/klaytn/blob/v1.10.2/consensus/istanbul/backend/snapshot.go#L208-L210

So, between the block number 3600*k and 3600*(k+1), the proposer refresh is skipped due to this line: 

https://github.com/klaytn/klaytn/blob/v1.10.2/consensus/istanbul/validator/weighted.go#L663-L666

Reply:
> I think "no staking info" is a rare error case.

Could you elaborate on this?

By https://github.com/klaytn/klaytn/blob/v1.10.2/reward/staking_manager.go#L130, 

Numbers between `k * 3600`  and `24k * 3600` hits that line, resulting in no invoke proposer list update.

Reply:
The numbers between `k*3600` and `24k*3600` would be converted to to `k'*86400` by `CalcStakingBlockNumber` before querying `GetStakingInfoOnStakingBlock`.

So, to summarize, if block number is 100000,
- snapshot.apply calls `Refresh()` with `blockNum` 92700 in the argument: https://github.com/klaytn/klaytn/blob/v1.10.2/consensus/istanbul/backend/snapshot.go#L208-L210
- `Refresh()` calls reward.GetStakingInfo(92701): https://github.com/klaytn/klaytn/blob/v1.10.2/consensus/istanbul/validator/weighted.go#L640
- `GetStakingInfo()` calls GetStakingInfoOnStakingBlock(86400): https://github.com/klaytn/klaytn/blob/v1.10.2/reward/staking_manager.go#L111
- `params.IsStakingUpdateInterval(stakingBlockNumber)` is true

Reply:
[I overlooked this](https://github.com/klaytn/klaytn/blob/v1.10.2/reward/staking_manager.go#L109). Sorry for the noise. Confirmed that shuffling is triggered by every interval of `proposerUpdateInterval`. Thank you.